controlpanel/api/auth0.py

```diff
@@ -0,0 +1,190 @@
+            get_token = authentication.GetToken(self.domain)
+
+            try:
+                token = get_token.client_credentials(
+                    self.client_id, self.client_secret, self.audience
+                )
+
+            except exceptions.Auth0Error as error:
+                raise Auth0Error(
+                    error, self.client_id, self.domain, self.audience
+                )
+
+            self._access_token = token["access_token"]
+
+        return self._access_token
+
+    def request(self, method, endpoint, **kwargs):
+        base_url = self.base_url
+        if not base_url.endswith('/'):
+            base_url = base_url + '/'
+        url = f"{base_url}{endpoint}"
+        response = requests.request(
+            method,
+            url,
+            headers={
+                "Content-Type": "application/json",
+                "Authorization": f"Bearer {self.access_token}",
+            },
+            **kwargs,
+        )
+        response.raise_for_status()
+
+        if response.text:
+            return response.json()
+
+        return {}
+
+    def get_all(self, endpoint, key=None, **kwargs):
+        items = []
+        total = None
+        if "params" not in kwargs:
+            kwargs["params"] = {}
+        if "page" in kwargs:
+            kwargs["params"]["page"] = kwargs.pop("page")
+        if "per_page" in kwargs:
+            kwargs["params"]["per_page"] = kwargs.pop("per_page")
+
+        if key is None:
+            key = endpoint
+
+        while True:
+            response = self.request("GET", endpoint, **kwargs)
+
+            if "total" not in response:
+                raise Auth0Error(f"get_all {endpoint}: Missing total")
+
+            if key not in response:
+                raise Auth0Error(f"get_all {endpoint}: Missing key {key}")
+
+            if total is None:
+                total = response["total"]
+
+            if total != response["total"]:
+                raise Auth0Error(f"get_all {endpoint}: Total changed")
+
+            items.extend(response[key])
+
+            if len(items) >= total:
+                break
+
+            if len(response[key]) < 1:
+                break
+
+            if "page" not in kwargs["params"]:
+                kwargs["params"]["page"] = 1
+
+            kwargs["params"]["page"] += 1
+
+        return items
+
+
+class AuthorizationAPI(APIClient):
+    base_url = settings.AUTH0["authorization_extension_url"]
+    audience = "urn:auth0-authz-api"
+
+    def get_group(self, group_name):
+        groups = [
+            group for group in self.get_all("groups") if group["name"] == group_name
+        ]
+
+        if len(groups) != 1:
+            return None
+
+        return groups[0]
+
+    def get_group_members(self, group_name):
+        group = self.get_group(group_name)
+        if group:
+            return self.get_all(
+                f'groups/{group["_id"]}/members', key="users", per_page=25
+            )
```

Comment:
#### 1d. Refactor Auth0 client module

> Lines 109 - 173

...and subclasses for the Authorization API, providing convenience methods wrapping API requests to manage group members...

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-4abd63b561b5260e074659dad215fa1cR176)